data/clean.ts

```diff
@@ -84,19 +76,107 @@ async function parseAndIndexGrades(): Promise<
         subjectIndex[subjectArea][catalogNumber] = [];
       }
       subjectIndex[subjectArea][catalogNumber].push(row);
+
+      const summaryKey = `${enrollmentTerm}|${subjectArea}|${catalogNumber}|${sectionNumber}`;
+      if (!summaryIndex[summaryKey]) {
+        summaryIndex[summaryKey] = [];
+      }
+      summaryIndex[summaryKey].push(row);
     });
 
   return new Promise((resolve) => {
     parser.on("end", () => {
-      resolve([rows, instructorIndex, subjectIndex]);
+      resolve([rows, instructorIndex, subjectIndex, summaryIndex]);
     });
   });
 }
 
+function calculateSummaries(summaryIndex: SummaryIndex): SummaryRow[] {
+  const GRADE_VALUES = {
+    "A+": 4,
+    A: 4,
+    "A-": 3.7,
+    "B+": 3.3,
+    B: 3,
+    "B-": 2.7,
+    "C+": 2.3,
+    C: 2,
+    "C-": 1.7,
+    "D+": 1.3,
+    D: 1,
+    "D-": 0.7,
+    F: 0,
+  };
+
+  return Object.entries(summaryIndex).map(([summaryKey, rows]) => {
+    let totalWithGrade = 0;
+    let totalGradeValue = 0;
+    let standardErrors = 0;
+
+    rows.forEach((row) => {
+      const gradeValue =
+        GRADE_VALUES[row.gradeOffered as keyof typeof GRADE_VALUES];
+      if (gradeValue >= 0) {
+        totalWithGrade += Number(row.gradeCount);
+        totalGradeValue += gradeValue * Number(row.gradeCount);
+      }
+    });
```

Comment:
Some of this data aggregation/transformation code can probably be written more concisely using `lodash` — e.g. we could have something like

```ts
const totalGradeValue = sumBy(
  rows,
  (row) => {
    return row.gradeCount * getGradeValue(row.gradeOffered)
  }
)
```

In general I prefer this more functional style since it makes thinking about what's changing/happening easier